src/optimizer/filter_combiner.cpp

```diff
@@ -987,4 +994,201 @@ ValueComparisonResult CompareValueInformation(ExpressionValueInformation &left,
 	}
 }
 
+void FilterCombiner::LookUpConjunctions(Expression *expr) {
+	if (expr->GetExpressionType() == ExpressionType::CONJUNCTION_OR) {
+		auto root_or = (BoundConjunctionExpression *)expr;
+		for (auto &or_to_push : ors_to_pushdown) {
+			if (or_to_push->root_or->Equals(root_or)) {
+				return;
+			}
+		}
+
+		auto or_to_push = make_unique<OrToPush>();
+		or_to_push->root_or = root_or;
+		or_to_push->cur_conjunction = root_or;
+		ors_to_pushdown.emplace_back(move(or_to_push));
+
+		BFSLookUpConjunctions(root_or);
+		return;
+	}
+
+	// Verify if the expression has a column already pushed down by other OR expression
+	VerifyOrsToPush(*expr);
+}
+
+void FilterCombiner::BFSLookUpConjunctions(BoundConjunctionExpression *conjunction) {
+	if (VerifyEarlyStopPushdown()) {
+		return;
+	}
+
+	vector<BoundConjunctionExpression *> conjunctions_to_visit;
+
+	for (auto &child : conjunction->children) {
+		switch (child->GetExpressionClass()) {
+		case ExpressionClass::BOUND_CONJUNCTION: {
+			auto child_conjunction = (BoundConjunctionExpression *)child.get();
+			conjunctions_to_visit.emplace_back(child_conjunction);
+			break;
+		}
+		case ExpressionClass::BOUND_COMPARISON: {
+			UpdateConjunctionFilter((BoundComparisonExpression *)child.get(), *child);
+			break;
+		}
+		default: {
+			SetEarlyStopPushdown(true);
+			return;
+		}
+		}
+	}
+
+	for (auto child_conjunction : conjunctions_to_visit) {
+		SetCurrentConjunction(child_conjunction);
+		// traverse child conjuction
+		BFSLookUpConjunctions(child_conjunction);
+	}
+}
+
+void FilterCombiner::VerifyOrsToPush(Expression &expr) {
+	if (expr.type == ExpressionType::BOUND_COLUMN_REF) {
+		auto &colref = (BoundColumnRefExpression &)expr;
+		for (auto &or_to_push : ors_to_pushdown) {
+			if (or_to_push->col_conjunction) {
+				// case we have the same column in AND expression, we should give priority to it
+				if (or_to_push->col_conjunction->column_ref->Equals(&colref)) {
+					or_to_push->stop_pushdown = true;
+				}
+			}
+		}
+	}
+	ExpressionIterator::EnumerateChildren(expr, [&](Expression &child) { VerifyOrsToPush(child); });
+}
+
+void FilterCombiner::UpdateConjunctionFilter(BoundComparisonExpression *comparison_expr, Expression &expr) {
+	if (expr.type == ExpressionType::BOUND_COLUMN_REF) {
+		bool left_is_scalar = comparison_expr->left->IsFoldable();
+		bool right_is_scalar = comparison_expr->right->IsFoldable();
+		// only support comparison with scalar
+		if (left_is_scalar || right_is_scalar) {
+			auto non_scalar_expr = left_is_scalar ? comparison_expr->right.get() : comparison_expr->left.get();
+			if (non_scalar_expr->GetExpressionType() == ExpressionType::BOUND_COLUMN_REF) {
+				UpdateFilterByColumn((BoundColumnRefExpression *)&expr, comparison_expr);
+				return;
+			}
+		}
+		SetEarlyStopPushdown(true);
+	}
+	ExpressionIterator::EnumerateChildren(expr,
```

Comment:
Should this iterate through the entire expression tree, even if the `SetEarlyStopPushdown` is triggered?

It also seems a bit odd to recursively iterate through the entire expression tree, e.g. what if I have something like `x=7 IS NULL`? Then `x` does not need to be equal to `7`, after all. Should this not only look at the elements that are actually contained in conjunctions?